### PR TITLE
ANOVA Functionality

### PR DESCRIPTION
**Pull request by frank113**
Resolves # .

This pull request will add ANOVA functionality. 

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [y] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [y] Read and understand the [Code of Conduct][code-of-conduct].
-   [y] Read and understood the [licensing terms][license].
-   [y] Searched for existing issues and pull requests **before** submitting this pull request.
-   [y] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [y] Rebased onto latest `develop`.
-   [y] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   Creates an `anova` function to perform one-way ANOVA given an array of continuous values and an array of factors. Support for other input methods will be added.
-   b
-   c

## Related Issues

> Does this pull request have any related issues?

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

### CODE REVIEW

File: lib/node_modules/@stdlib/math/stats/anova1/README.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+<!-- eslint no-undef: "error" -->
+
+
+</section>
+
+<!-- /.examples -->
+
+<section class="links">
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/README.md]: Please add 

```
[mdn-array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array

[mdn-typed-array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays

```

so that the Markdown reference links are not missing their definitions and are rendered correctly.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
+# One Way ANOVA
+
+> Perform One-Way ANOVA on given data
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/README.md]: Let's change the description to `Perform a one-way analysis of variance.`

File: lib/node_modules/@stdlib/math/stats/anova1/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+## Usage
+
+```javascript
+var anova1 = require( '@stdlib/math/stats/anova1' );
+```
+
+### anova1( x, factor\[, ...params]\[, opts] )
+
+For an [array][mdn-array] or [typed array][mdn-typed-array] of numeric values `x` and an [array][mdn-array] of classifications `factor`, one-way anova1 is performed. The hypotheses are given as follows:
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/README.md]: Let's revert this undesired change that was introduced when renaming the package: `s/anova1/anova`.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+$$H_{0}: \mu_{1} = \mu_{2} = \dots = \mu_{k}$$
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/README.md]: We have a special way of writing formulas so that we can later automatically render the equations using MathJax. For this one, we could do 

```
<!-- <equation class="equation" label="eq:hypotheses" align="center" raw="\begin{align*} H_{0}:& \; \mu_{1} = \mu_{2} = \dots = \mu_{k} \\ H_{a}:& \; \text{at least one} \; \mu_{i} \; \text{not equal to the others} \end{align*}" alt="Hypotheses of ANOVA"> -->

<!-- </equation> -->

```

File: lib/node_modules/@stdlib/math/stats/anova1/examples/index.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+var ANOVA = require('../lib/make_anova.js');
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/examples/index.js]: Let's rename `ANOVA` to `anova1` here.

File: lib/node_modules/@stdlib/math/stats/lib/index.js [hunk 6]
```diff
@@ -243,7 +252,6 @@ setReadOnly( statistics, 'ztest', require( '@stdlib/math/stats/ztest' ) );
 */
 setReadOnly( statistics, 'ztest2', require( '@stdlib/math/stats/ztest2' ) );
 
-
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/lib/index.js]: Can we please revert this change? Before section headers (`// MODULES //`, `// VARIABLES //`, `// FUNCTIONS //`, `// MAIN //`, `// EXPORTS //`, `// TESTS //`), we always use two empty lines.

File: lib/node_modules/@stdlib/math/stats/anova1/test/test.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.equal( typeof anova1, 'function', 'main export is a function' );
+	t.end();
+});
+
+// Check to make sure X is an array --> Copied from chi2gof
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/test/test.js]: Let's remove these comments. The test descriptions suffice.

File: lib/node_modules/@stdlib/math/stats/anova1/package.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "@stdlib/math/stats/anova1",
+  "version": "0.0.0",
+  "description": "Examine 1-way ANOVA relationships",
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/package.json]: Let's change the description to `Perform a one-way analysis of variance.`

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+var setReadOnly = require( '@stdlib/utils/define-read-only-property' );
+var cdf = require( '@stdlib/math/base/dists/f/cdf' );
+var copy = require( '@stdlib/utils/copy' );
+var defaults = require( './defaults.json' );
+var validate = require( './validate.js' );
+var uniqueElems = require('./unique_elems.js');
+var meanTable = require('./mean_table.js');
+var prettyPrint = require( './print.js' );
+
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Before the main function in a file, we always place a `// MAIN // ` header (except for the case when there is nothing coming before the main function).

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		throw new RangeError( 'length of x must be greater than or equal to 2.');
+	}
+
+	// Check if factor is an array
+	if ( !isArray( factor )) {
+		throw new TypeError(' factor argument must be an array' );
+	}
+
+	if (factor.length <= 1) {
+		throw new RangeError('Second argument must have at least 2 elements');
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: I think we can remove this if-clause since we check in L83 for two unique elements.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	treats = uniqueElems(factor);
+	if (!treats) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: We can remove this if-clause. `uniqueElems` should always return an array, even if it may be empty.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+* Make the ANOVA Object.
+*
+* @param {NumericArray} x - Measured values
+* @param {Array} factor - Array of treatments
+* @param {Options} [options] - function options
+* @param {number} [options.alpha=0.05] - significance level
+* @param {boolean} [options.decision=false] - Indicate whether we reject the null hypothesis
+* @returns {Object} Returns the ANOVA Structure
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: In JSDoc comments, we document it when the functions throw errors. Before the `@returns` annotation, we should add:

```
* @throws {TypeError} options argument must be an object
* @throws {TypeError} must provide valid options
* @throws {TypeError} `x` must be a numeric array
* @throws {TypeError} `factor` must be an array
* @throws {RangeError} `factor` must have at least two unique elements
* @throws {RangeError} length of `x` must be greater than or equal to 2
* @throws {RangeError} `x` and `factor` must have the same length
```

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+	// Now to make the out object
+	out = {};
+	setReadOnly(out, 'treatment_df', treats.length - 1);
+	setReadOnly(out, 'treatment_sum_squares', ssTreatment);
+	setReadOnly(out, 'treatment_mean_sum_squares', meanSumSqTreat);
+	setReadOnly(out, 'error_df', x.length - treats.length);
+	setReadOnly(out, 'error_sum_squares', sumSqError);
+	setReadOnly(out, 'error_mean_sum_squares', meanSumSqError);
+	setReadOnly(out, 'F', fScore);
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Let's rename this to `statistic` to match the other `stats` packages. We will need to update the names used inside of the `print` method so that it continues to work.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+	setReadOnly(out, 'P_Value', pVal);
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Change from `P_Value` to `pValue` to match the other statistical test packages.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+// MODULES //
+
+var incrstdev = require( '@stdlib/math/stats/incr/stdev' );
+var uniqueElems = require('./unique_elems.js');
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: Let's add `// MAIN //` header here.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+/**
+* Computes the vector of means for each treatment.
+*
+* @param {NumericArray} x - Measured values
+* @param {Array} factor - Array of treatments
+* @returns {{}} destination array
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: `@returns {{}}` should be `@returns {Object}`. The description `destination array` also needs to be updated.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+	var factorCount;
+	var meanTable;
+	var j;
+	var i;
+	var accumulator;
+	var k;
+	var newMean;
+
+
+	if (x.length !== factor.length) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: It's fine to remove this check given since this function is only called inside of `make_anova.js` (where we have already established that the arrays have the same length) and won't be publicly available.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+* Computes the vector of means for each treatment.
+*
+* @param {NumericArray} x - Measured values
+* @param {Array} factor - Array of treatments
+* @returns {{}} destination array
+*/
+function meanTable(x, factor) {
+	var treatments;
+	var factorCount;
+	var meanTable;
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: Can we rename the `meanTable` variable? As-is, it has the same name as the function, which can be confusing.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/index.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+/**
+* One-Way Analysis of Variance.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/index.js]: `One-way analysis of variance.`

File: lib/node_modules/@stdlib/math/stats/anova1/test/test.js [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+tape( 'the function produces expected values on simple inputs', function test( t ) {
+	var x;
+	var f;
+	var out;
+
+	x = [3, 4, 5, 6, 2, 5, 10, 12, 8, 10];
+	f = ['frank', 'philipp', 'nugent', 'frank', 'philipp', 'nugent', 'frank', 'philipp', 'nugent', 'frank'];
+	out = anova1(x, f);
+
+	t.equal(out.treatment_df, 2);
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/test/test.js]: Where do these values come from? We should add a comment, e.g. `// Tested against R:`. Better yet would be to have a script to generate the R fixtures automatically. But we can add that later.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+	fScore = meanSumSqTreat / meanSumSqError;
+
+	// Now get the p-value
+	numDf = treats.length - 1;
+	denDf = x.length - treats.length;
+	pVal = 1 - cdf(fScore, numDf, denDf);
+
+	// Now to make the out object
+	out = {};
+	setReadOnly(out, 'treatment_df', treats.length - 1);
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Can we avoid underscores in property names? We could use names like `treatSumSq` and `errMeanSq`? Not so sure about the degrees of freedom. What do you think, @frank113 and @kgryte?

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Awesome, Frank! Thanks for all your work. I left some comments on things we should resolve before merging in this pull request.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 21 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: I would take inspiration from MATLAB (https://www.mathworks.com/help/stats/anova1.html), including in terms of the output data structures.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 1 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: Resolved.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 2 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: Resolved.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 3 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: I am unsure what you mean here. I am taking this to mean the removal of the optional parameters that we discussed after the 200 staff meeting on Tuesday. I will ask about this when we meet tomorrow.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 4 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: Resolved. Out of curiosity, does the markdown format always revert to MathJax?

File: lib/node_modules/@stdlib/math/stats/anova1/examples/index.js (hunk 5 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/examples/index.js]: Resolved.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/index.js (hunk 19 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/index.js]: Resolved.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 9 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: I placed it in `make_anova.js` below the description and above the `anova1` function.

File: lib/node_modules/@stdlib/math/stats/anova1/test/test.js (hunk 7 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/test/test.js]: Resolved.

File: lib/node_modules/@stdlib/math/stats/lib/index.js (hunk 6 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/lib/index.js]: Resolved in deep_copy, ztest2, test and stats/index.js

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 4 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: Yes, as MathJax is what we use for server-side equation rendering. See https://github.com/stdlib-js/stdlib/blob/4b962089f5b1c7fbd82fb45d50ec534b6cb410ad/lib/node_modules/%40stdlib/_tools/utils/tex-equation-to-svg/lib/factory.js#L23.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 3 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: I believe @Planeshifter is referring to the description:

> ... classifications `factor`, one-way **anova** is performed.

File: lib/node_modules/@stdlib/math/stats/anova1/package.json (hunk 8 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/package.json]: Resolved.

File: lib/node_modules/@stdlib/math/stats/anova1/README.md (hunk 4 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/README.md]: Thanks!

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 10 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: I updated the RangeError text and removed the L83 conditional, as I am not quite sure what I wanted to accomplish with the former. The latter checks for at least two unique elements, L83 was a relic from when the function could return null.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 11 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Resolved in previous comment.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 12 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Resolved. Must the text match the error messages generated in the `throw new`?

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 13 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Resolved in `make_anova.js` and `print.js`

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 14 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Resolved in `make_anova.js` and `print.js`

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js (hunk 15 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: Added. I also added a space above `// MODULES //` to keep in convention with the two empty spaces.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js (hunk 16 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: Both resolved.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js (hunk 17 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: Ok. This may be useful for future development yet for now I agree with the sentiment. Removed.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 12 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: No, not necessarily. There should, however, be some level of agreement such that people reading the code do not see conflicting messages about an error's intent.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js (hunk 18 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: `tableOfMeans`

File: lib/node_modules/@stdlib/math/stats/anova1/test/test.js (hunk 20 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/test/test.js]: These values came from R testing side by side, forgot to add the R testing fixtures. Will do at a later date.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 21 above)
Comment (frank113) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: @kgryte The MATLAB looks much cleaner than what I have; far more readable. I will update the naming to mirror it. For `Prob>F` could we simply replace it with `p`?

File: lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js (hunk 18 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/mean_table.js]: You probably don't have to be as descriptive. It would be enough to name the variable `out`. The meaning of the variable is readily available from the function docs, and, as it stands now, one has to read until function end to know what value is returned from the function.

File: lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js (hunk 21 above)
Comment (Planeshifter) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Instead of `Prob>F`, let's use `pValue` to be consistent with the other stats packages.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/stats/anova1/lib/make_anova.js]: Yes, or `pValue`, whichever is the convention currently used my @Planeshifter elsewhere.
